bigquery/src/main/java/org/apache/iceberg/gcp/bigquery/BigQueryMetastoreCatalog.java

```diff
@@ -0,0 +1,400 @@
+    Dataset builder = new Dataset();
+    DatasetReference datasetReference = toDatasetReference(namespace);
+    builder.setLocation(this.location);
+    builder.setDatasetReference(datasetReference);
+    builder.setExternalCatalogDatasetOptions(
+        BigQueryMetastoreUtils.createExternalCatalogDatasetOptions(
+            datasetReference.getDatasetId(), metadata));
+
+    client.createDataset(builder);
+  }
+
+  /**
+   * Since this catalog only supports one-level namespaces, it always returns an empty list unless
+   * passed an empty namespace to list all namespaces within the catalog.
+   */
+  @Override
+  public List<Namespace> listNamespaces(Namespace namespace) {
+    if (namespace.levels().length != 0) {
+      // BQMS does not support namespaces under database or tables, returns empty.
+      // It is called when dropping a namespace to make sure it's empty (listTables is called as
+      // well), returns empty to unblock deletion.
+      return ImmutableList.of();
+    }
+    return client.listDatasets(projectId).stream()
+        .map(BigQueryMetastoreCatalog::getNamespace)
+        .collect(ImmutableList.toImmutableList());
+  }
+
+  @Override
+  public boolean dropNamespace(Namespace namespace) {
+    try {
+      client.deleteDataset(toDatasetReference(namespace));
+      // We don't delete the data folder for safety, which aligns with Hive Metastore's default
+      // behavior.
+      // We can support database or catalog level config controlling file deletion in the future.
+      return true;
+    } catch (NoSuchNamespaceException e) {
+      return false;
+    }
+  }
+
+  @Override
+  public boolean setProperties(Namespace namespace, Map<String, String> properties) {
+    Dataset dataset = client.getDataset(toDatasetReference(namespace));
+
+    ExternalCatalogDatasetOptions existingOptions = dataset.getExternalCatalogDatasetOptions();
+    Map<String, String> existingParameters =
+        existingOptions != null ? existingOptions.getParameters() : null;
+
+    Map<String, String> newParameters = Maps.newHashMap();
+    if (existingParameters != null) {
+      newParameters.putAll(existingParameters);
+    }
+    newParameters.putAll(properties);
+
+    if (Objects.equals(existingParameters, newParameters)) {
+      // No change in parameters detected
+      return false;
+    }
+
+    client.setDatasetParameters(toDatasetReference(namespace), properties);
+    return true;
+  }
+
+  @Override
+  public boolean removeProperties(Namespace namespace, Set<String> properties) {
+    client.removeDatasetParameters(toDatasetReference(namespace), properties);
+    return true;
+  }
+
+  @SuppressWarnings("FormatStringAnnotation")
+  @Override
+  public Map<String, String> loadNamespaceMetadata(Namespace namespace) {
+    try {
+      return getMetadata(client.getDataset(toDatasetReference(namespace)));
+    } catch (IllegalArgumentException e) {
```

Comment:
it seems a bit weird to handle IAE like this and then throw a NSNE

Reply:
If Namespace string does not support from BigQuery it will throw IAE. It is NSNE for user because they can not create in first place.

Reply:
so the IAE is something that the BigQuery client is throwing? I'm a bit worried that if the IAE happens for a different reason that it ends up as a NSNE at the user